### DISCUSSION TITLE
A implementation for CAD

### DISCUSSION THREAD

**Discussion by papyDoctor** [datamole-ai/gomez#31] [General]
Hello,
Following our discussion [here](https://github.com/datamole-ai/gomez/issues/8), I've done a small crate from my project where all the rendering and ui are removed to concentrate on the constraint problem, you can find it [here](https://github.com/papyDoctor/test-gomez). Was a lot of work to bind my data structure to yours.
In the example, I've four points on which I can apply constraint like segment parallel, horizontal, vertical, point fixed, point fixed on x axis, point fixed on y axis...
If I've exactly 8 equations then it's a determined system and the solver find every time the solution.
But if I remove a constraint (undetermined system), sometimes it converges, sometime not.
I can arrange to have always determined system but it will be not easy I think.

What do you think about this? 
And do you plan to use Fear soon?

Olivier

**Comment by pnevyk** [comment]
Thanks!

> Was a lot of work to bind my data structure to yours.

If you have any thoughts on how this experience could be improved, I would love to hear that. Were there any specific obstacles? Or would you imagine the API structured differently?

> But if I remove a constraint (undetermined system), sometimes it converges, sometime not.

Just to be sure, success/failure  is deterministic right? So for a specific problem with specific initial values it always either succeeds or fails.

Likely not the problem, but I suggest to zero the `rx` vector at the beginning of `eval` (I believe you could use `rx.fill(0)`). It's not guaranteed that the passed vector is all zeros (the implementation might use various temp vectors), so zeroing it manually makes sure that there are no leftover non-zero values causing the solver to think that it didn't converge. (I have an idea how to support underdetermined systems more conveniently API-wise, but until then, this is necessary.)

In order to debug the problematic cases, can you install a log-compatible logger (e.g., [env_logger](https://docs.rs/env_logger/0.10.1/env_logger/)) and set it to output debug level logs? It will print a lot of things, but it should give a better idea about what is going on under the hood. If it is possible, you could also paste the logs here so I can try to make sense out of them.

> And do you plan to use Fear soon?

Yes, this migration is planned for the upcoming weeks in my job, so unless something unexpected happens, I believe that it will be released soon-ish.
